numpy/core/src/common/simd/avx512/conversion.h

```diff
@@ -90,6 +90,96 @@ NPY_FINLINE npyv_u32x2 npyv_expand_u32_u16(npyv_u16 data)
+    __m512i  r = _mm512_packs_epi16(p1, p2);
+    return npyv_cvt_b8_u8(_mm512_permutexvar_epi32(idx, r));
+#else
+    const __m512i idx = npyv__setr_epi32(
+        /* a */0, 4, 1, 5,   /* b */2, 6, 3, 7,
+        /* c */8, 12, 9, 13, /* d */10, 14, 11, 15);
+    __m256i ta, tb, tc, td, lo, hi;
+    NPYV__PACK_LO_HI(32, npyv_cvt_u32_b32(a), ta);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u32_b32(b), tb);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u32_b32(c), tc);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u32_b32(d), td);
+    __m256i p1 = _mm256_packs_epi16(ta, tb);
+    __m256i p2 = _mm256_packs_epi16(tc, td);
+    __m512i  r = npyv512_combine_si256(p1, p2);
+    return npyv_cvt_b8_u8(_mm512_permutexvar_epi32(idx, r));
+#endif
+}
+
+NPY_FINLINE npyv_b8
+npyv_pack_b8_b64(npyv_b64 a, npyv_b64 b, npyv_b64 c, npyv_b64 d,
+                 npyv_b64 e, npyv_b64 f, npyv_b64 g, npyv_b64 h) {
+#ifdef NPY_HAVE_AVX512BW
+    const __m512i idx = npyv__setr_epi16(
+        /* a */0, 8, 16, 24,  /* b */1, 9, 17, 25,  /* c */2, 10, 18, 26,
+        /* d */3, 11, 19, 27, /* e */4, 12, 20, 28, /* f */5, 13, 21, 29, 
+        /* g */6, 14, 22, 30, /* h */7, 15, 23, 31);
+    __m512i p1 = _mm512_packs_epi32(npyv_cvt_u64_b64(a), npyv_cvt_u64_b64(b));
+    __m512i p2 = _mm512_packs_epi32(npyv_cvt_u64_b64(c), npyv_cvt_u64_b64(d));
+    __m512i p3 = _mm512_packs_epi32(npyv_cvt_u64_b64(e), npyv_cvt_u64_b64(f));
+    __m512i p4 = _mm512_packs_epi32(npyv_cvt_u64_b64(g), npyv_cvt_u64_b64(h));
+    __m512i p1_2 = _mm512_packs_epi32(p1, p2);
+    __m512i p2_2 = _mm512_packs_epi32(p3, p4);
+    __m512i r = _mm512_packs_epi16(p1_2, p2_2);
+    return npyv_cvt_b8_u8(_mm512_permutexvar_epi16(idx, r));
+#else
+    const __m256i idx = _mm256_setr_epi32(
+        /* a|c|e|g */0, 4, 1, 5, /* b|d|f|h */2, 6, 3, 7);
+    const __m512i idx2 = npyv__setr_epi64(
+        /* a,b,c,d */0, 2, 1, 3, /* e,f,g,h */4, 6, 5, 7);
+    __m256i ta, tb, tc, td, te, tf, tg, th, lo, hi;
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(a), ta);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(b), tb);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(c), tc);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(d), td);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(e), te);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(f), tf);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(g), tg);
+    NPYV__PACK_LO_HI(32, npyv_cvt_u64_b64(h), th);
+    __m256i p1 = _mm256_permutevar8x32_epi32(_mm256_packs_epi32(ta, tb), idx);
+    __m256i p2 = _mm256_permutevar8x32_epi32(_mm256_packs_epi32(tc, td), idx);
+    __m256i p3 = _mm256_permutevar8x32_epi32(_mm256_packs_epi32(te, tf), idx);
+    __m256i p4 = _mm256_permutevar8x32_epi32(_mm256_packs_epi32(tg, th), idx);
+    __m256i p1_2 = _mm256_packs_epi16(p1, p2);
+    __m256i p2_2 = _mm256_packs_epi16(p3, p4);
+    __m512i r = npyv512_combine_si256(p1_2, p2_2);
+    return npyv_cvt_b8_u8(_mm512_permutexvar_epi64(idx2, r));
+#endif
+}
+
+#undef NPYV__PACK_LO_HI
```

Comment:
I used your code to implement the function `npyv_pack_b8_b64` (AVX512F) with a few changes (see below):

```
    __m512i ai = npyv_cvt_u32_b32(a);
    __m512i bi = npyv_cvt_u32_b32(b);
    __m512i ci = npyv_cvt_u32_b32(c);
    __m512i di = npyv_cvt_u32_b32(d);
    __m512i ab = npyv512_packs_epi32(ai, bi);
    __m512i cd = npyv512_packs_epi32(ci, di);
+    __m512i abcd = npyv512_packs_epi16(ab, cd);  <-- added
+    return npyv512_shuffle_odd32(abcd);          <-- added
-    return return npyv_pack_b16(ab, cd);         <-- removed
```

and I noticed some slowdown comparing your solution vs mine:

```
AVX512F:
export NPY_DISABLE_CPU_FEATURES="AVX512_SKX"

+      8.98±0.3μs      10.4±0.09μs     1.16  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float32'>)
+      8.99±0.4μs      10.4±0.07μs     1.15  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint32'>)
+     9.73±0.05μs      11.2±0.08μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int32'>)
+     9.78±0.07μs      11.3±0.09μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint32'>)
+      9.83±0.1μs       11.3±0.1μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float32'>)
+     9.77±0.07μs      11.2±0.08μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int32'>)
+      9.81±0.1μs       11.3±0.1μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float32'>)
+     9.79±0.05μs       11.2±0.2μs     1.15  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint32'>)
+      9.11±0.4μs      10.4±0.08μs     1.14  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int32'>)
-     19.1±0.06μs      14.3±0.07μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)
-     19.1±0.09μs      14.3±0.07μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)
-     19.2±0.08μs      14.3±0.05μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)
-     19.2±0.06μs      14.3±0.09μs     0.75  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)
-     19.3±0.07μs      14.4±0.01μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)
-      19.3±0.5μs      14.3±0.05μs     0.74  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>)
```

What is happening (IMO) is that your code does some conversions that can be avoided. Keeping the data as __m256i and only converting it back to __m512i at the end of all computation seems to be better. For this case, in particular, I am keeping my code but with some small changes. If you want me to change it, let me know.


See below the improvements after changing the code and comparing it with the initial solution of this PR.

#### AVX512F:

```
export NPY_DISABLE_CPU_FEATURES="AVX512_SKX"

-      20.5±0.2μs       18.1±0.9μs     0.88  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.float64'>)                                                                      
-      20.4±0.4μs         17.9±1μs     0.87  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.int64'>)                                                                        
-        21.6±2μs         16.5±2μs     0.76  bench_ufunc.CustomComparison.time_less_than_binary(<class 'numpy.uint64'>)                                                                       
-     19.2±0.08μs      12.8±0.01μs     0.67  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int64'>)                                                                       
-     19.2±0.06μs      12.8±0.01μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.float64'>)                                                                     
-     19.2±0.08μs      12.8±0.03μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int64'>)                                                                       
-     19.2±0.07μs      12.7±0.02μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint64'>)                                                                      
-      19.3±0.3μs      12.8±0.03μs     0.66  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint64'>)                                                                      
-      19.7±0.4μs      12.8±0.06μs     0.65  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.float64'>) 
``` 


#### AVX512BW:

```
unset NPY_DISABLE_CPU_FEATURES

-     5.42±0.09μs      5.15±0.09μs     0.95  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.int16'>)
-     5.38±0.06μs      5.06±0.02μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar2(<class 'numpy.uint16'>)
-     5.41±0.04μs      5.08±0.03μs     0.94  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.uint16'>)
-     5.45±0.06μs      5.05±0.03μs     0.93  bench_ufunc.CustomComparison.time_less_than_scalar1(<class 'numpy.int16'>)
```

Reply:
> What is happening (IMO) is that your code does some conversions that can be avoided.

I agree, there're extra shuffles in my version(AVX512F only), feel free to modified but keep using  `_mm512_kunpackd, _mm512_kunpackw, _mm512_kunpackb` to avoid unnecessary mask conversion.